Initial version of manual bind to arrow function

vjeux opened the pull request:
This converts the manual bind to an arrow property initializer function.

It works for the base case `this.method = this.method.bind(this);`, two other patterns that I've seen in the codebase that should be codemodded as well in a subsequent version

- `(this: any).method = this.method.bind(this)`
- `const self = (this: any); self.method = this.method.bind(this)`

Test Plan:
npm t

I haven't tried it on the codebase yet, just want to get early feedback

keyanzhang commented on transforms/manual-bind-to-arrow.js:
```diff
@@ -0,0 +1,85 @@
+      )) {
+        return false;
+      }
+
```
I think you can replace the above with
```js
  root
    .find(j.MethodDefinition, {
      kind: 'constructor',
    })
    .find(j.AssignmentExpression, {
      left: {
        type: 'MemberExpression',
        object: {
          type: 'ThisExpression',
        },
        property: {
          type: 'Identifier',
        },
      },
      right: {
        type: 'CallExpression',
        callee: {
          type: 'MemberExpression',
          property: {
            type: 'Identifier',
            name: 'bind',
          },
          object: {
            type: 'MemberExpression',
            object: {
              type: 'ThisExpression',
            },
            property: {
              type: 'Identifier',
            },
          },
        },
      },
    })
    .filter(path => 
      path.node.left.property.name === path.node.right.callee.object.property.name
    )
```

vjeux replied:
Thanks for the trick! In the new version I've added conditionals inside of that logic so I'm not really sure we can do it as you mentioned.

Also, for the method lookup, I need the reference to that node, so I'd need to traverse it up anyway.

keyanzhang commented on transforms/manual-bind-to-arrow.js:
```diff
@@ -0,0 +1,85 @@
+      return true;
+    })
+    .map(path => path.parentPath)
+    .replaceWith(path => null)
```
If I recall correctly it might be better to do 
```js
.forEach(path => j(path).remove())
```
here. `.replaceWith(path => null)` leaves a semicolon here: https://astexplorer.net/#/kHE62wfbBC

vjeux replied:
Thanks! Done

cpojer commented on transforms/manual-bind-to-arrow.js:
```diff
@@ -0,0 +1,85 @@
+      }
+
+      if (!(
+        path.node.left.type === 'MemberExpression' &&
```
this whole thing might benefit from some variable extraction like `left = path.node.left, right = path.node.right, callee = right.callee` etc. etc.

vjeux replied:
I like having it in one giant if, I can just comment stuff out and console.log things in the process. I'm not sure it would be really more readable or anything splitting them out in variables.

vjeux commented:
- Added support for:
  - `(this: any).method = this.method.bind(this);`
  - `const self: any = this; self.method = self.method.bind(this);`
- Now removes the constructor when there's no content left but a super call.
- It bails out early if there's nothing to transform.
- Added unit tests for all those.

I'm going to run this on www now and see how it looks. All the features I want are now supported, there are probably going to be things uncovered trying it out in the real codebase.

vjeux commented:
- Adding support for async functions
- Use the babylon parser to avoid issues with different line location from flow parser

vjeux commented:
Fixing lint error on one of the test that prevented travis to be green :)

keyanzhang commented:
Since this has already been battle tested let's merge this? 😃
